jac-uk/admin
# Feature/2005  User Management - Functionality Improvements

## What's included?
Closes #2005 

- [x] Create a `users` collection in Firestore for pagination and search.
- [x] Write a node script to populate data from the authentication database into `users` collection.
- [x] Update UI.
- [x] Update sign-in process.
- [x] Add a background function `onUserDelete` to delete the user in the authentication database.
- [x] Add a background function `onUserUpdate` to update the user information (e.g. displayName, email, disabled) in the authentication database.
- [x] Add a background function `onRoleUpdate` to mark users with this role as changed.
- [x] Update the contact dropdown to get the user list from the `users` collection.
- [x] Remove hard-coded email addresses.
- [x] Create a `userInvitations` collection to invite new users.

Note: This PR relies on the changes in [digital-platform: Feature/2005 user management improvements](https://github.com/jac-uk/digital-platform/pull/910).

### Development plan

#### Sign in
1. `auth.onAuthStateChanged(user)` is triggered.
2. Dispatch `auth/setCurrentUser` action.
3. Check if the email is valid.
4. Check if the user document exists in Firestore.
    - No (new user)
        1. Check if there is a user invitation document with the status `pending` in Firestore.
        2. Create a user document in Firestore.
        3. Update the user role information in the custom claims.
        4. Mark the status of the user invitation as `completed`.
5. Ensure the Microsoft email is verified. 
6. Refresh the token using `auth.currentUser.getIdToken(true)` to get the latest custom claims.
7. Get custom claims using `user.getIdTokenResult()`
8. Set the store (`bindCurrentUser`, `setRolePermissions`)

#### A user’s role is changed
Mark `role.isChanged` as `true` in the user document.

[digital-platform] `onUserUpdate`
1. Update role permissions in the custom claims.
5. Mark `role.isChanged` as `false` in the user document.

Note: The currently logged-in users can get new permissions after refreshing the page.

#### Role permissions are changed
Update role permissions in the role document.

[digital-platform] `onRoleUpdate`
1. Get users with this role and mark `role.isChanged` as true.
2. Trigger `onUserUpdate` and update role permissions in the custom claims.

Note: The currently logged-in users can get new permissions after refreshing the page.

#### `role.isChanged` is changed to `true`
Show pop-up to refresh the page.

#### New admin users
1. Create a user invitation for new users.
2. New users log in using their Google/Microsoft account.
3. Check if a user invitation has been created. If a user invitation exists in the Firestore, the user can log in to the platform with the given user role.


## Who should test?
✅ Product owner
✅ Developers

## How to test?

Preview URL: https://jac-admin-develop--pr2054-feature-2005-user-ma-xvbo9tuh.web.app

1. Check if you can sign in to the admin site without problems.

2. Go to `Users` page:
    - Check if all admin users are listed in the table.
    - Check if you can search for a specific user.
    - Check if the pagination of the user table works.
    - Check if you can enable/disable/delete a user.
    - Check if you can change the role.

3. Go to `Roles` page:
    - Check if you can see all roles.
    - Check if you can change the permissions of a role.
    - Check if you can create a new role.

4. Go to `User Invitations` page:
    - Check if you can create a user invitation with an JAC email address and user role.
    - Check if you can use the email to sign in. 


## Risk - how likely is this to impact other areas?
🟠 Medium risk - this does change code that is shared with other areas

## Additional context


## Related permissions
Permissions have been added:
- Can read users `u7`

---
PREVIEW:DEVELOP
_can be OFF, DEVELOP or STAGING_
